### PR TITLE
Add SMC calls forwarding framework

### PR DESCRIPTION
**Pull request by malus-brandywine**
Added:
- a framework for forwarding SMC calls to Secure Monitor,
- SiP Service entry in the framework.

### CODE REVIEW

**Comment by malus-brandywine** [comment]
Design note: smc.c file provided an SMC router `handle_smc()`, which called individual Service handlers.

Now there's the place in the file, separated by  conditional compilation,  to list handlers of Services that require "physical" access (vs. emulation) to EL3 components. At the moment, only SiP Service handler was declared.

There's also  common _default service handler_  (unconditional forwarder) which can be assigned to be a handler for any of mentioned Services.

Also, by default,  the Service handlers are initialized with NULL in the library, so vmm application should explicitly pick the default handler or implement own one.

File: src/arch/aarch64/smc.c [hunk 1]
```diff
@@ -32,6 +33,29 @@ typedef enum {
...
+#if defined(CONFIG_ALLOW_SMC_CALLS)
+/* Service handlers for services that require "physical" access to EL3 components
+   (vs emulation) are listed here.
+
+   'Active' handler is a pointer to fucntion that actually will be invoked;
+   by default active handlers are set to NULL. Application should assign to an active handler
+   wheither a default handler that unconditionally forwards calls to Secure Monitor:
+
+   extern handle_service_type handle_xxx;
+   handle_service_type handle_xxx = default_handle_service;
```
Comment (Ivan-Velickovic) [inline src/arch/aarch64/smc.c:45]: In my opinion the user of the library should not have to do this, and instead just have a wrapper for setting the handler.

File: src/arch/aarch64/smc.c [hunk 2]
```diff
@@ -109,10 +133,44 @@ bool handle_smc(size_t vcpu_id, uint32_t hsr)
...
 }
+
+
+/* Default handler of calls to a Service (group of Functions) running at Secure Monitor level;
+  forwards all the calls without applying any policy
+*/
+bool default_handle_service(uint64_t vcpu_id, seL4_UserContext *regs, uint64_t fn_number)
+{
+
+    seL4_CPtr smc_cap = SMC_CAP;
```
Comment (Ivan-Velickovic) [inline src/arch/aarch64/smc.c:155]: While it isn't the current case for the rest of the library, I would like the library to be agnostic to the seL4 environment it is in, this means that it should be up to the user to provide the library with what capability to use for SMC calls.

If it's not too hard for you, it'd be good if specifying the SMC cap was part of the API.

File: src/arch/aarch64/smc.h [hunk 3]
```diff
@@ -20,3 +20,12 @@ void smc_set_return_value(seL4_UserContext *u, uint64_t val);
 
 /* Gets the value of x1-x6 */
 uint64_t smc_get_arg(seL4_UserContext *u, uint64_t arg);
+
+/* Default handler of calls to a SMC Service */
+bool default_handle_service(uint64_t vcpu_id, seL4_UserContext *regs, uint64_t fn_number);
```
Comment (Ivan-Velickovic) [inline src/arch/aarch64/smc.h:25]: nit: `vcpu_id` should be `size_t`.

File: src/arch/aarch64/smc.c (hunk 1 above)
Comment (Ivan-Velickovic) [reply src/arch/aarch64/smc.c:45]: Also thank you for writing comments for how to use the API.

Comment (malus-brandywine) [reply src/arch/aarch64/smc.c:45]: > In my opinion the user of the library should not have to do this, and instead just have a wrapper for setting the handler.

Agreed, that would be easier to use.

File: src/arch/aarch64/smc.c (hunk 2 above)
Comment (malus-brandywine) [reply src/arch/aarch64/smc.c:155]: Right, good catch, I'll change it.

File: src/arch/aarch64/smc.c (hunk 1 above)
Comment (malus-brandywine) [reply src/arch/aarch64/smc.c]: API changed

File: src/arch/aarch64/smc.h (hunk 3 above)
Comment (malus-brandywine) [reply src/arch/aarch64/smc.h]: Done

File: src/arch/aarch64/smc.c (hunk 2 above)
Comment (malus-brandywine) [reply src/arch/aarch64/smc.c]: Done

**Comment by malus-brandywine** [comment]
Number of added names (functions, vars) somewhat noticeable (and may increase eventually over time), so I added prefixes `smc_` to everything newly added, otherwise it makes a zoo.

File: src/arch/aarch64/smc.h [hunk 4]
```diff
@@ -20,3 +20,22 @@ void smc_set_return_value(seL4_UserContext *u, uint64_t val);
...
+bool smc_set_cap(seL4_CPtr smcccap);
+
+/* Default handler of calls to a SMC Service */
+bool smc_default_handle_service(size_t vcpu_id, seL4_UserContext *regs, uint64_t fn_number);
+
+/* Service handler type */
+typedef bool (* smc_handle_service_type)(size_t, seL4_UserContext *, uint64_t);
+
+/* Default handler of SiP service calls */
+bool smc_default_handle_sip(size_t vcpu_id, seL4_UserContext *regs, uint64_t fn_number);
```
[the remark this answers is not in the document]
Comment (malus-brandywine) [reply src/arch/aarch64/smc.h]: Precisely! Fixed.

**Comment by malus-brandywine** [comment]
Ping. Ping.

**Comment by Ivan-Velickovic** [comment]
> Ping. Ping.

Will get to it in the next couple days, thanks for the reminder.

**Comment by Ivan-Velickovic** [comment]
> Will get to it in the next couple days, thanks for the reminder.

Well looks like a couple days turned into a couple months, sorry about that.

I've made https://github.com/au-ts/libvmm/pull/119 which takes your changes and rebases, I made some minor changes as well.

Thanks for the PR and I'll try be more responsive next time :)

**Comment by malus-brandywine** [comment]
Thank you!

**Pull request closed without merging**